### PR TITLE
Feature/platfowner/genesis sharding

### PR DESCRIPTION
**Pull request by platfowner**
Summary:
- Support sharding config with genesis_sharding.json
- Support custom genesis configs with GENESIS_CONFIGS_DIR env variable
- Refactor genesis block generation
- Let default genesis config when no custom config is given (minimize code duplication)
- Add getJsObject(), setJsObject() util functions
- Fix bugs in genesis functions / rules / owners config 
- Add tests for sharding

Result:
- We can run shard nodes using e.g.:
```
GENESIS_CONFIGS_DIR=blockchain/afan_shard NUM_VALIDATORS=1 ACCOUNT_INDEX=0 HOSTING_ENV=local DEBUG=false node client/index.js
```

#56

### CODE REVIEW

File: blockchain/block.js [hunk 1]
```diff
@@ -329,17 +236,14 @@ class Block {
     // This is a temporary fix for the genesis block. Code should be modified after
     // genesis block broadcasting feature is implemented.
     const ownerAccount = GenesisAccounts.owner;
-    const timestamp = GenesisAccounts.timestamp;
+    const timestamp = Date.now();
```
Comment (minsulee2) [inline blockchain/block.js:239]: Finally! We've got now here :) What about genesisTime or something?

File: chain-util.js [hunk 2]
```diff
@@ -87,12 +87,50 @@ class ChainUtil {
...
     return (formatted.startsWith('/') ? '' : '/') + formatted;
   }
 
+  static getJsObject(obj, path) {
+    if (!ChainUtil.isArray(path)) {
+      return null;
+    }
+    let ref = obj;
+    for (let i = 0; i < path.length; i++) {
+      const key = String(path[i]);
```
Comment (minsulee2) [inline chain-util.js:102]: What about using static toString() function here?

File: chain-util.js [hunk 3]
```diff
@@ -87,12 +87,50 @@ class ChainUtil {
...
+    }
+    if (!ChainUtil.isDict(obj)) {
+      return false;
+    }
+    if (path.length == 0) {
+      return false;
+    }
+    let ref = obj;
+    for (let i = 0; i < path.length - 1; i++) {
+      const key = String(path[i]);
```
Comment (minsulee2) [inline chain-util.js:123]: ditto

File: chain-util.js [hunk 4]
```diff
@@ -87,12 +87,50 @@ class ChainUtil {
...
+      if (!ChainUtil.isDict(ref[key])) {
+        ref[key] = {};
+      }
+      ref = ref[key];
+    }
+    const key = String(path[path.length - 1]);
```
Comment (minsulee2) [inline chain-util.js:129]: ditto

File: blockchain/block.js (hunk 1 above)
Comment (platfowner) [reply blockchain/block.js]: Done. Note that the timestamp is used for both block timestamp and transaction timestamp.

File: chain-util.js (hunk 2 above)
Comment (platfowner) [reply chain-util.js]: Done

File: chain-util.js (hunk 3 above)
Comment (platfowner) [reply chain-util.js]: ditto

File: chain-util.js (hunk 4 above)
Comment (platfowner) [reply chain-util.js]: ditto

**Comment by platfowner** [comment]
Comments addressed. PTAL @minsulee2

File: blockchain/block.js [hunk 5]
```diff
@@ -329,18 +236,15 @@ class Block {
     // This is a temporary fix for the genesis block. Code should be modified after
     // genesis block broadcasting feature is implemented.
     const ownerAccount = GenesisAccounts.owner;
-    const timestamp = GenesisAccounts.timestamp;
+    const genesisTime = Date.now();
```
[the remark this answers is not in the document]
Comment (minsulee2) [reply blockchain/block.js]: Oh, I missed it too. It should be updated at some point tho.

**Review by minsulee2 — APPROVED** [review]
LGTM!

**Comment by platfowner** [comment]
Thanks for the quick review. Now merging..

**Pull request merged**